### PR TITLE
[SPARK-49663][SQL] Enable RTRIM suggestions in collation expressions

### PR DESCRIPTION
**Pull request by jovanpavl-db**
### What changes were proposed in this pull request?
Add RTRIM collation suggestion for collation expressions in case of the wrong name.

### Why are the changes needed?
Better customer experience. 

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
Added tests in CollationFactorySuite.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@jovanpavl-db Please, fix the coding style issue:
```
[error] /__w/spark/spark/common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala: illegal start of simple expression: Token(RPAREN,),4621,))
```

**Review by MaxGekk — COMMENTED** [review]
@stefankandic @uros-db @stevomitric Could you review this PR, please.

**Review by uros-db — APPROVED** [review]
this is fine, although just a note: we may need to re-visit this again in the future in order to improve the overall suggestions algorithm

for example, `SR_CI_RTRIM_AI` is a valid collation, while `SR_CI_TRIM_AI` is not - although I doubt that the suggestions will be very accurate in this case (and many others)

**Review by MaxGekk — COMMENTED** [review]
Does it make sense to use the levenshtein distance for suggestions like we do in `UNRESOLVED_FIELD.WITH_SUGGESTION`, see
https://github.com/apache/spark/blob/bb15eb7b91ab775bdb84b6b17353a706794b122d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala#L75-L97

**Comment by jovanpavl-db** [comment]
> Does it make sense to use the Levenshtein distance for suggestions like we do in `UNRESOLVED_FIELD.WITH_SUGGESTION`, see
> 
> https://github.com/apache/spark/blob/bb15eb7b91ab775bdb84b6b17353a706794b122d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala#L75-L97

I went through this implementation and it seems useful, however it can be quite problematic in our case to do suggestion on specifiers (we currently want 100% match on them because they are usually 2 chars long), for example if someone puts _C you can suggest _CI, _CS and even some UNICODE collations like ci. For every wrong specifier we can have multiple suggestions so it can really explode. That's why I think we (currently) stayed out of business of doing suggestions on specifiers but rather whole collation. We can definitely rethink this approach in the future at least for the some of specifiers like trim.

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala [hunk 1]
```diff
@@ -106,21 +106,27 @@ class CollationFactorySuite extends AnyFunSuite with Matchers { // scalastyle:ig
       ("UTF8_BINARY_AI","UTF8_BINARY"),
       ("UTF8_BINARY_CI_AI","UTF8_BINARY"),
       ("UTF8_BINARY_AI_CI","UTF8_BINARY"),
+      ("UTF8_BINARY_AI_RTRIM", "UTF8_BINARY_RTRIM"),
+      ("UTF8_BINARY_CI_RTRIM","UTF8_BINARY_RTRIM"),
+      ("UTF8_BINARY_AI_CI_RTRIM","UTF8_BINARY_RTRIM"),
```
Comment (MaxGekk) [inline common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:111]: let's be consistent and use spaces:
```suggestion
      ("UTF8_BINARY_CI_RTRIM", "UTF8_BINARY_RTRIM"),
      ("UTF8_BINARY_AI_CI_RTRIM", "UTF8_BINARY_RTRIM"),
```

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala [hunk 2]
```diff
@@ -106,21 +106,27 @@ class CollationFactorySuite extends AnyFunSuite with Matchers { // scalastyle:ig
...
       ("UTF8_BS","UTF8_LCASE"),
       ("BINARY_UTF8","ar_SAU"),
       ("UTF8_BINARY_A","UTF8_BINARY"),
       ("UNICODE_X","UNICODE"),
       ("UNICODE_CI_X","UNICODE"),
       ("UNICODE_LCASE_X","UNICODE"),
+      ("UNICODE_RTRIM_LCASE_X","UNICODE"),
```
Comment (MaxGekk) [inline common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:118]:
```suggestion
      ("UNICODE_RTRIM_LCASE_X", "UNICODE"),
```

**Review by MaxGekk — COMMENTED** [review]
lgtm except of a couple comments.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @jovanpavl-db and @uros-db for your review.

**Pull request closed without merging**